WordPress/WordPressTest/TaggedManagedObjectIDTests.swift

```diff
@@ -0,0 +1,89 @@
+        let context = contextManager.mainContext
+        let post = PostBuilder(context).with(title: "Test post").build()
+        let id = try TaggedManagedObjectID(unsaved: post)
+
+        let result = try context.existingObject(with: id)
+        XCTAssertEqual(result.postTitle, "Test post")
+    }
+
+    func testQueryUnsavedUsingDifferentContext() throws {
+        let post = PostBuilder(contextManager.mainContext).build()
+        let id = try TaggedManagedObjectID(unsaved: post)
+
+        let newContext = contextManager.newDerivedContext()
+
+        let expectation = expectation(description: "The query should fail because the model is not saved")
+        do {
+            _ = try newContext.existingObject(with: id)
+        } catch {
+            expectation.fulfill()
+        }
+        wait(for: [expectation], timeout: 0.1)
+    }
+
+    func testEqutable() throws {
+        let post = PostBuilder(contextManager.mainContext).with(title: "Test post").build()
+        let unsaveID = try TaggedManagedObjectID(unsaved: post)
+        try contextManager.mainContext.save()
+        let savedID = TaggedManagedObjectID(saved: post)
+
+        XCTAssertEqual(unsaveID, savedID)
+        XCTAssertEqual(TaggedManagedObjectID(saved: post), savedID)
+    }
+
+    // This is not really a test. More like a demostration of how to workaround Swift compiler's covariance support.
+    func testCovariance() throws {
+        let post = PostBuilder(contextManager.mainContext).with(author: "WordPress.com").build()
+        try contextManager.mainContext.save()
+
+        let postID: TaggedManagedObjectID<Post> = .init(saved: post)
+        let abstractPostID: TaggedManagedObjectID<AbstractPost> = .init(saved: post)
+
+        // This line does not compile without the second `author(of:in:)` overload.
+        try XCTAssertEqual(author(of: postID, in: contextManager.mainContext), "WordPress.com")
+        try XCTAssertEqual(author(of: abstractPostID, in: contextManager.mainContext), "WordPress.com")
+    }
+
+    // This function does not accept `TaggedManagedObjectID<Post>` argument, because `TaggedManagedObjectID<Post>` is
+    // not a subtype of `TaggedManagedObjectID<AbstractPost>` even though `Post` is a subtype of `AbstractPost`.
+    //
+    // But sometimes, we do want to declare a function that accepts object id of all `AbstractPost` types. In those cases,
+    // we can declare them in another way: see the function below.
+    private func author(of id: TaggedManagedObjectID<AbstractPost>, in context: NSManagedObjectContext) throws -> String? {
+        try context.existingObject(with: id).author
+    }
+
+    // This function accepts all `TaggedManagedObjectID` instances that are created from all `AbstractPost` types.
+    private func author<T: AbstractPost>(of id: TaggedManagedObjectID<T>, in context: NSManagedObjectContext) throws -> String? {
+        try context.existingObject(with: id).author
+    }
```

Comment:
Maybe it's me not being familiar with the Swift compiler support, but I had to read the test and comments twice to understand what was going on 😄 😳

Here's a suggestion. I think if a bit more info had been provided at the start, everything would have been clear.

```suggestion
    // MARK: – Covariance workaround demo
    
    // Covariance is the ability to use a more derived type (a subclass) in place of a less derived type (a superclass).
    //
    // When we "wrap" an `NSManagedObject` in a `TaggedManagedObjectID<Model>`, we lose type-hierarchy information. 
    // That means the compiler will not accept a `TaggedManagedObjectID<Post>` instance for a parameter defined as `TaggedManagedObjectID<AbstractPost>`
    // even though `Post` is a subtype of `AbstractPost`.
    //
    // Below is a demonstration of how to work around this limitation by using a function overload.
    
    // If this test compiles, then the overload definition (see below) works.
    func testCovariance() throws {
        let post = PostBuilder(contextManager.mainContext).with(author: "WordPress.com").build()
        try contextManager.mainContext.save()

        let postID: TaggedManagedObjectID<Post> = .init(saved: post)
        let abstractPostID: TaggedManagedObjectID<AbstractPost> = .init(saved: post)

        // This line does not compile without the second `author(of:in:)` overload.
        try XCTAssertEqual(author(of: postID, in: contextManager.mainContext), "WordPress.com")
        try XCTAssertEqual(author(of: abstractPostID, in: contextManager.mainContext), "WordPress.com")
    }

    // This function does not accept `TaggedManagedObjectID<Post>` argument, because `TaggedManagedObjectID<Post>` is
    // not a subtype of `TaggedManagedObjectID<AbstractPost>` even though `Post` is a subtype of `AbstractPost`.
    //
    // But sometimes, we do want to declare a function that accepts object id of all `AbstractPost` types. In those cases,
    // we can declare them in another way: see the function below.
    private func author(of id: TaggedManagedObjectID<AbstractPost>, in context: NSManagedObjectContext) throws -> String? {
        try context.existingObject(with: id).author
    }

    // This function accepts all `TaggedManagedObjectID` instances that are created from all `AbstractPost` types.
    private func author<T: AbstractPost>(of id: TaggedManagedObjectID<T>, in context: NSManagedObjectContext) throws -> String? {
        try context.existingObject(with: id).author
    }
```

Reply:
Thanks for the suggestions! I have taken it and added more information. Let me know what you think. See https://github.com/wordpress-mobile/WordPress-iOS/pull/21381/commits/270484926837a7ad4520f727099fbd0701404493